"ReplaceWithMetadata"-Feature

Elfangor93 opened the pull request:
Ich habe vor einiger Zeit einen Thread im joomla.de-Forum gemacht, wie und ob man Bilder in der JoomGallery nach Aufnahmedatum (Exif-Aufnahmedatum) sortieren lassen kann. Nach der Rückmeldung von Mr.Music habe ich ein Plugin geschrieben, welches das Exif-Datum ausliest und das Upload-Datum des Bildes in der Datenbank überscheibt. Jedoch war es mir damit nicht möglich Meldungen dem User auszugeben...
[exifsort.zip](https://github.com/JoomGallery/JoomGallery/files/2014096/exifsort.zip)
Ich habe von Alwin darauf diese Änderung am Core erhalten, mit welchem sich sämtliche Bild-Daten durch Meta-Daten ersetzen lassen. -> Danke Alwin, ich finde dies ein super Feature!

Ich habe dann seine Änderungen um ein Meldesystem erweitert, damit der User informiert wird, falls Meta-Daten in seinem Bild fehlen oder sonst etwas beim Uploadprozess bezüglich "Ersetzen durch Meta-Daten" nicht geklappt hat. Diese Meldungen werden mit "Warnung" deklariert und immer ausgegeben, wenn "Ersetzen durch Meta-Daten" aktiviert ist und Probleme damit aufgetreten sind, unabhängig ob Debugging aktiviert wurde oder nicht.

Ich bin blutiger Anfänger in php und GitHub. Ich hoffe, ich richte mit diesen Änderungen nicht zu viel Chaos in eurem Projekt an. Ich finde die JoomGallery jedoch eine super Extension und bin stolz, wenn ich (hoffentlich) etwas nützliches zu diesem Projekt beitragen konnte...

Lg Manuel

Elfangor93 committed:
Language Strings für "ReplaceWithMetadata"-Feature

Mehrere Langugae Strings hinzugefügt, welche für das "ReplaceWithMetadata"-Feature benötigt werden. Mehrheitlich für Debug- oder Warnhinweisen...

Elfangor93 committed:
Backend Language Strings für "ReplaceWithMetadata"-Feature

Mehrere Langugae Strings hinzugefügt, welche für das "ReplaceWithMetadata"-Feature benötigt werden. Mehrheitlich Text für die neuen Einstellungen im Konfigurations-Manager (Tab "Ersetzen durch Meta-Daten").

Elfangor93 committed:
Neues Tab im Konfigurations-Manager

Generelle Einstellungen - Ersetzen durch Meta-Daten hinzugefügt
Einstellungsmöglichkeit, ob gewisse Bildinformationen aus den Metadaten des entsprechenden Bildes geladen werden sollen.
(Ich hoffe das stimmt so. Änderung vorgenommen durch Mr.Music)

Elfangor93 committed:
Neues Tab im Konfigurations-Manager

Generelle Einstellungen - Ersetzen durch Meta-Daten hinzugefügt
Einstellungsmöglichkeit, ob gewisse Bildinformationen aus den Metadaten des entsprechenden Bildes geladen werden sollen.
(Ich hoffe das stimmt so. Änderung vorgenommen durch Mr.Music)

Elfangor93 committed:
Tabellenfelder für neues Tab im Konfigurations-Manager

Generelle Einstellungen - Ersetzen durch Meta-Daten hinzugefügt
Neue Tabellenfelder, in denen die Einstellungen aus dem Tab Ersetzen durch Meta-Daten gespeichert werden
(Ich hoffe das stimmt so. Änderung vorgenommen durch Mr.Music)

Elfangor93 committed:
Neue Tabellenfelder für Ersetzen durch Meta-Daten Tab 

Konfigurations-Manager: Generelle Einstellungen - Ersetzen durch Meta-Daten hinzugefügt
Tabellenfelder durch update hinzufügen
(Ich hoffe das stimmt so. Änderung vorgenommen durch Mr.Music)

Elfangor93 committed:
Fixing Redirect

Änderung, dass Redirect im Ajax-Upload nur stattfindet, wenn der Upload erfolgreich war und keine Warnungen oder Debug-Messages ausgegeben werden sollen.
Dies geschieht über die neue ifRedirect-Methode in der class JoomUpload (administrator/.../helpers/upload.php)

Elfangor93 committed:
Verschiedene neue Methoden (Ersetzen durch Meta-Daten)

Die Funktion Ersetzen durch Meta-Daten wird in den upload-prozess integriert. Dies geschieht hauptsächlich durch die neue Methoden "getOverrideValues" und "readMetaData". (Ich hoffe das stimmt so. Änderung vorgenommen durch Mr.Music)
Zusätzlich wurde $_warningoutput hinzugefügt, um Rückmeldungen ausgeben zu können, auch wenn Debug ausgeschaltet ist. Die Methode getDebugOutput wurde erweitert und als Ausgabe für alle Upload-Varianten gesetzt (echo $this->_debugoutput ersetzt). Die Language Strings für IPTC und EXIF Daten würden integriert und die Nummer-Werte aus den Settings mit der Methode getMetaName aufgeschlüsselt.
... und viele weitere kleine Änderungen...

MrMusic commented:
Vielen Dank für den Pull Request. Bei einem kurzen Test hat alles gut geklappt.
Vielleicht wäre eine Option im Konfigurations-Manager zum Ein- und Ausschalten der Meldungen bei fehlenden Meta-Daten sinnvoll? Evtl. möchte nicht jeder auf seiner Seite die Meldungen gezeigt bekommen.

Elfangor93 commented:
Könnte man noch, ja. Wie kann man eine Konfiguration hinzufügen?
Auch die Texte der Meldungen finde ich noch nicht ganz gelungen...

Elfangor93 committed:
Strings für neue Konfiguration

Strings für Konfiguration>>Generelle Einstellungen>>Ersetzen durch Meta-Data>>Zeige Warnungen

Elfangor93 committed:
Logik, um Warnung ein-/auszuschalten

logik eingefügt, damit Konfiguration berücksichtigt wird (ob Warnungen angezeigt werden sollen oder nicht)

Elfangor93 committed:
neue Konfiguration zu DB hinzufügen

neuer Eintrag in der DB für die Konfiguration>>Generelle Einstellungen>>Ersetzen durch Meta-Daten>>Zeige Warnungen

Elfangor93 committed:
Anzeige der neuen Konfiguration

Konfiguration>>Generelle Einstellungen>>Ersetzen durch Meta-Daten>>Zeige Warnungen

Elfangor93 commented:
Konfiguration hinzugefügt. Es kann nun ausgewählt werden, ob Warnungen, falls vorhanden angezeigt werden sollen oder nicht. Und, ob mit den Warnungen ein Hinweis für den Uploader augegeben werden soll, wie er das Problem lösen kann.

Elfangor93 commented:
Bei meinem Test zu den letzten Änderungen ist mir aufgefallen, dass der Upload (wenn Meta-Ersetzungen aktiv sind) bei nicht .jpg-Formaten als fehlgeschlagen angezeigt werden.
Test mit IPTC (Beschreibung wird durch IPTC-Beschreibung ersetzt:
![grafik](https://user-images.githubusercontent.com/39154009/41925715-aafa6ff6-796d-11e8-9583-f0d9bc3221cf.png)
Test mit EXIF (Datum wird durch Exif-Aufnahmedatum ersetzt):
![grafik](https://user-images.githubusercontent.com/39154009/41925812-f5a72eea-796d-11e8-9a20-3f558360ed58.png)
Die Bilder werden trotzdem hochgeladen und erscheinen auch in der Galerie. Die Ersetzung hat jedoch nicht geklappt.
Vermutung: die verwendeten php-Funktionen (exif_read_data, iptcparse) kommen nur mit .jpg klar. Da müsste man vieleicht noch eine exception machen...

Elfangor93 commented:
Laut php.net wird der Parameter imageinfo in getimagesize(), welcher für iptcparse() gebraucht, nur für JFIF-Files (jpg) unterstützt. (Quelle: https://secure.php.net/manual/en/function.getimagesize.php)

Und exif_read_data() kann laut php.net nur exif-daten aus  JPEG und TIFF-files lesen. (Quelle: https://secure.php.net/manual/en/function.exif-read-data.php)

Da muss man sich entscheiden, für welche Bild-Formate diese Funktion "ReplaceWithMetadata" überhaupt verfügbar sein soll. Sollen z.B Metadaten aus png-files (Chunk - http://dev.exiv2.org/projects/exiv2/wiki/The_Metadata_in_PNG_files) auch berücksichtigt werden? Gibt es noch andere Metadaten?

Vielleicht wäre das was: https://github.com/gomoob/php-metadata-extractor
Oder das: https://github.com/JamesHeinrich/getID3/
Oder das (falls nur jpeg untertsützt werden soll): http://www.ozhiker.com/electronics/pjmt/

MrMusic commented:
Erstmal vielen Dank für die zusätzliche Konfiguration.
Meine Meinung:
Da die Funktionen zum lesen der Exif- und IPTC-Daten nur mit .jpg-Dateien zuverlässig funktionieren, sollten wir uns darauf beschränken. Ich glaube nicht, dass ein großer Bedarf nach anderen Dateiformaten oder anderen Meta-Daten (z.B. xmp) besteht. Falls doch könnte man das immer noch implementieren.
Um eine php-Warning zu verhindern, müsste man jetzt noch einen Check auf die richtige Dateierweiterung einbauen. Beispiel:
```
      // check for jpg extension
      $tag = strtolower(JFile::getExt($readfile));
      if($tag == 'jpg' || $tag == 'jpeg' || $tag == 'jpe')
      {
        $exif_array = exif_read_data($readfile);
      }
```

PS: Kleiner Schreibfehler in der en-GB.com_joomgallery.ini:
COM_JOOMGALLERY_UPLOAD_OUTPUT_WARNING_REPLACEIMGDATE="- There is no exif-date in the **mate**-data...

Elfangor93 commented:
Ich werde noch einen Hinweis in der Konfiguration platzieren, dass "Ersetzen durch Meta-Daten" nur mit jpeg-Files möglich ist.
Dann wäre nur noch die Sache mit dem JPG abzuklären. Warum funktionieren .jpg-Dateien, aber .JPG-Dateien nicht?
Das jpeg-Bildformat kommt ja mit mehreren Endungen daher (in Gross und Kleinschreibung):
![grafik](https://user-images.githubusercontent.com/39154009/42038907-655dbac2-7aec-11e8-9c6a-b9c05a7c3779.png)
Quelle: https://de.wikipedia.org/wiki/JPEG_File_Interchange_Format
Da müsste man noch einen Workaround finden. Eventuell vor dem Auslesen der Metadaten Files umbenennen nach .jpg?

Elfangor93 committed:
Warnungs-String für nicht-jpeg-formate

Warnungstext, der ausgegeben wird, wenn Metadaten eines nicht jpeg-files ausgelesen werden sollten

Elfangor93 committed:
Überprüfung auf file-format vor dem ersetzen

Logik, damit für nicht jpeg-formate statt dem ersetzen der Metadaten eine Warnung ausgegeben wird.

Elfangor93 commented:
Bei gewissen Bildern kommt immernoch der Error "Uploud faild", obwohl soweit ich das sehe alles korrekt durchgelaufen ist.
Finde den Ursprung des Fehlers nicht...

MrMusic commented:
>Dann wäre nur noch die Sache mit dem JPG abzuklären. Warum funktionieren .jpg-Dateien, aber .JPG-Dateien nicht?

Ich habe keine Probleme mit dem Upload, egal ob Groß- oder Kleinschreibung.

1. Ich persönlich würde die Warnung bei falschem Dateiformat nur dann anzeigen, wenn dies in der Konfiguration eingestellt ist.
Beispiel:
`
if (!($tag == 'jpg' || $tag == 'jpeg' || $tag == 'jpe' || $tag == 'jfif'))
  {
      if($this->_config->get('jg_replaceshowwarning') > 0)
      {
        $this->_warningoutput .= JText::_('COM_JOOMGALLERY_UPLOAD_OUTPUT_WARNING_WRONGFILEFORMAT').': .'.$tag.'<br />';
        $metaWarning = true;
   }
}
`

2. Die Funktion getIfRedirect in der upload.php scheint mir nirgens verwendet zu werden

Hinweis zu Code-Style:
Die Einrückungen im Code der JG sind bisher immer mit Leerzeichen gemacht, nicht mit Tabulator.

Erftralle commented:
Vielen Dank für deinen Beitrag. Weitere Codekommentare folgen ...

Erftralle commented on administrator/components/com_joomgallery/sql/install.mysql.utf8.sql:
```diff
@@ -97,6 +97,13 @@ CREATE TABLE IF NOT EXISTS `#__joomgallery_config` (
   `jg_checkupdate` int(1) NOT NULL,
   `jg_filenamewithjs` int(1) NOT NULL,
   `jg_filenamereplace` text NOT NULL,
+  `jg_replaceimgtitle` text NOT NULL,
```
Warum wurde als Datentyp 'text' verwendet? Ein Integer-Typ wäre doch ausreichend. Zudem sollten weiter unten in dieser Datei beim Einfügen eines Datensatzes für die Konfiguration Standardwerte für diese neuen Optionen berücksichtigt werden.

Erftralle commented on administrator/components/com_joomgallery/sql/install.mysql.utf8.sql:
```diff
@@ -384,7 +391,7 @@ CREATE TABLE IF NOT EXISTS `#__joomgallery_config` (
 
 CREATE TABLE IF NOT EXISTS `#__joomgallery_countstop` (
   `cspicid` int(11) NOT NULL default 0,
-  `csip` varchar(45) NOT NULL default '',
```
Mit dieser Änderung würdest du den bereits "gemergten" #154 wieder zunichte machen.

Erftralle commented on administrator/components/com_joomgallery/sql/updates/mysql/3.4.0.sql:
```diff
@@ -1 +1,7 @@
-ALTER TABLE `#__joomgallery_countstop` MODIFY `csip` VARCHAR(45) NOT NULL DEFAULT '';
```
Mit dieser Änderung würdest du den bereits "gemergten" #154 wieder zunichte machen.

Erftralle commented on administrator/components/com_joomgallery/sql/updates/mysql/3.4.0.sql:
```diff
@@ -1 +1,7 @@
-ALTER TABLE `#__joomgallery_countstop` MODIFY `csip` VARCHAR(45) NOT NULL DEFAULT '';
+ALTER TABLE `#__joomgallery_config` ADD `jg_replaceimgtitle` text NOT NULL AFTER `jg_filenamereplace`;
```
Hier fehlt ein Standardwert, sonst sind die Optionen nach einem Update der Galerie undefiniert.

Erftralle commented on administrator/components/com_joomgallery/tables/joomgalleryconfig.php:
```diff
@@ -40,6 +40,13 @@ class TableJoomgalleryConfig extends JTable
   var $jg_checkupdate;
   var $jg_filenamewithjs;
   var $jg_filenamereplace;
+  var $jg_replaceimgtitle;
```
Die neuen Optionen sollten auch in der **_administrator\components\com_joomgallery\helpers\config.php_** eingetragen werden.

Erftralle commented on administrator/language/en-GB/en-GB.com_joomgallery.ini:
```diff
@@ -1563,3 +1563,78 @@ COM_JOOMGALLERY_HLPIFO_TEAM_CONTRIBUTORS="The most active contributors to the ma
 ;-------
 COM_JOOMGALLERY_UPLOAD_OUTPUT_MAX_ALLOWED_FILESIZE="Error: Max allowed file size is %s byte"
+; 
+;------------------------------------------------------------------------------------
```
Zeilen 1567 bis 1583 sind doppelt vorhanden.

Erftralle replied:
Diese Konstanten gehören zudem auch zu #106 und haben mit diesem nichts zu tun.

Erftralle commented on administrator/components/com_joomgallery/models/fields/fineuploader.php:
```diff
@@ -275,12 +275,12 @@ classes: {
       }
       <?php if(!$isMini): ?>
       uploader.requestParams.imgtext = jQuery('#<?php echo $prefix; ?>imgtext').val();
-      uploader.requestParams.imgauthor = jQuery('#<?php echo $prefix; ?>imgauthor').val();
```
Diese Änderung gehört eigentlich zu #149 und hat mit diesem "Pull Request" doch nichts zu tun, oder? Zudem wäre er unvollständig.

Erftralle commented on administrator/components/com_joomgallery/models/fields/fineuploader.php:
```diff
@@ -275,12 +275,12 @@ classes: {
       uploader.requestParams.published = jQuery('#<?php echo $prefix; ?>published').val();
       <?php   else: ?>
       uploader.requestParams.published = jQuery('#<?php echo $prefix; ?>published0').prop('checked') ? 0 : 1;
+      uploader.requestParams.imgauthor = jQuery('#<?php echo $prefix; ?>imgauthor').val();
```
Diese Änderung gehört eigentlich zu #149 und hat mit diesem "Pull Request" doch nichts zu tun, oder? Zudem wäre er unvollständig.

Erftralle commented on administrator/components/com_joomgallery/models/fields/fineuploader.php:
```diff
@@ -167,8 +167,8 @@ classes: {
             var redirect = '<?php echo $redirect; ?>';
-            if(uploader.fileCount == 0 && redirect != '') {
-              // Redirect only if all file uploads were successful
+            if(uploader.fileCount == 0 && redirect != '' && response.ifRedirect) {
```
"response.ifRedirect" wird nirgendswo im JoomGallery Code gesetzt.

Elfangor93 replied:
gefixt. siehe weiter unten...

Erftralle commented on administrator/components/com_joomgallery/helpers/upload.php:
```diff
@@ -62,7 +63,14 @@ class JoomUpload extends JObject
    *
    * @var string
    */
-  protected $_debugoutput = '';
```
Warum wurde die Sichtbarkeit hier geändert? "protected" wäre doch nach wie vor ausreichend.

Erftralle commented on administrator/components/com_joomgallery/helpers/upload.php:
```diff
@@ -62,7 +63,14 @@ class JoomUpload extends JObject
+   *
+   * @var string
+   */
+  public $_warningoutput = '';
```
"protected" wäre auch hier ausreichend.

Erftralle commented on administrator/components/com_joomgallery/helpers/upload.php:
```diff
@@ -62,7 +63,14 @@ class JoomUpload extends JObject
+  public $_debugoutput = '';
+
+  /**
+   * Holds warning which acure during the the upload procedure
```
Rechtschreibung: ... occur ...

Erftralle commented on administrator/components/com_joomgallery/helpers/upload.php:
```diff
@@ -407,6 +453,27 @@ protected function uploadSingles()
       }*/
 
+      // check for overriding with meta data
+      $overridevalues = array(
```
Soweit ich das gesehen habe, wird der nachfolgende Code für die verschiedenen Uploadmethoden immer wieder in gleicher Weise aufgerufen. Das schreit doch nach einer Methode.

MrMusic commented:
@Erftralle: Vielen Dank für deine Kommentare

Das Problem mit den Auswirkungen auf die anderen Pull Requests liegt wahrscheinlich daran, dass ich als "Grundlage" nicht den aktuellen 3.4.0-Branch von Github verwendet habe. 👎  
Deshalb waren die geänderten Dateien nicht auf dem richtigen Stand. Da hätte ich eigentlich vorher dran denken müssen.
Ich sehe jetzt zwei Möglichkeiten:
1. Entweder wir übernehmen die bereits gemergten Sachen in diesen Pull Request
2. Wir erstellen in Github einen neuen Branch, bauen dort das neue Replace-Feature ein (wie man es ja eigentlich machten sollte) und nehmen auch dort die weiteren Code-Verbesserungen vor.
Diese Möglichkeit würde ich persönlich bevorzugen.

@19leunam93: Wenn du mit Nr. 2 einverstanden bist, würde ich bei mir einen neuen Branch "ReplaceWithMetada" anlegen um den dann als Grundlage für einen neuen Pull Rquest zu verwenden.

Elfangor93 commented:
@MrMusic: Nr. 2 passt für mich...

MrMusic commented:
>@MrMusic: Nr. 2 passt für mich...

Danke!
Ich habe mal einen neuen Branch und einen neuen PR erstellt. Diesmal hoffentlich ohne Konflikte :-)

Elfangor93 commented:
Ich habe dir in deinem Fork die Sache mit dem Redirect nach dem Ajax-Upload eingereicht. Diese Änderungen beziehen sich auf die Beanstandung durch Erftralle. Es hat tatsächlich noch etwas gefehlt...
